dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala

```diff
@@ -16,345 +16,413 @@
+   * set the parameter of these engines correctly.
+   *
+   * The method can set parameters of multi-thread engines, verify spark conf values of an
+   * existing spark context.
    */
-  def invokeAndWait[T](tasks: Seq[() => T], timeout: Duration = Duration.Inf): Seq[T] = {
-    tasks.map(task => Future {
-      try {
-        task()
-      } catch {
-        case t : Throwable =>
-            logger.error("Error: " + ExceptionUtils.getStackTrace(t))
-            throw t
-      }
-    }(context)).map(future => {
-      Await.result(future, timeout)
-    })
+  def init: Unit = this.synchronized {
+    if (localMode) {
+      // The physical core number should have been initialized by env variable in bigdl.sh
+      setNodeAndCore(1, physicalCoreNumber)
+    } else {
+      logger.info("Auto detect executor number and executor cores number")
+      val (nExecutor, executorCores) = sparkExecutorAndCore(forceCheck = true).get
+      logger.info(s"Executor number is $nExecutor and executor cores number is $executorCores")
+      setNodeAndCore(nExecutor, executorCores)
+      checkSparkContext
+    }
   }
 
-  def invokeAndWait2[T](tasks: Seq[() => T], timeout: Long = Long.MaxValue,
-    timeUnit: TimeUnit = TimeUnit.NANOSECONDS):
-    scala.collection.mutable.Buffer[java.util.concurrent.Future[T]] = {
-    val callables = tasks.map(task => new Callable[T] {
-      override def call(): T = {
-        try {
-          task()
-        } catch {
-          case t : Throwable =>
-            logger.error("Error: " + ExceptionUtils.getStackTrace(t))
-            throw t
-        }
-      }
-    })
-    threadPool.invokeAll(callables.asJava, timeout, timeUnit).asScala
+  private val logger = Logger.getLogger(getClass)
+  private val singletonCounter: AtomicInteger = new AtomicInteger(0)
+  private var localMode: Boolean = {
+    val env = System.getenv("LOCAL_MODE")
+    if(env == null) {
+      false
+    } else {
+      true
+    }
   }
-
-  def invoke2[T](tasks: Seq[() => T]): Seq[java.util.concurrent.Future[T]] = {
-    tasks.map(task => new Callable[T] {
-      override def call(): T = {
-        task()
-      }
-    }).map(threadPool.submit(_))
+  private var physicalCoreNumber = {
+    if (!localMode) {
+      -1
+    } else {
+      getCoreNumberFromEnv
+    }
   }
+  private var nodeNum: Int = -1
 
-  /**
-   * Invoke a batch of tasks
-   *
-   * @param tasks
-   */
-  def invoke[T](tasks: Seq[() => T]): Seq[Future[T]] = {
-    tasks.map(task => Future {
-      try {
-        task()
-      } catch {
-        case t : Throwable =>
-          logger.error("Error: " + ExceptionUtils.getStackTrace(t))
-          throw t
-      }
-    }(context))
-  }
+  private val NOT_INIT_ERROR =
+    "Do you call Engine.init? See more at " +
+      "https://github.com/intel-analytics/BigDL/wiki/Programming-Guide#engine" +
+      "On executor side, you need to call Engine.setNodeAndCore before."
```

Comment:
User need to call Engine.setNodeAndCore to init node/core/thread pool if they want to use them on Executor. But this should be handled well by bigdl component. Remove it?